mopidy_alsamixer/mixer.py

```diff
@@ -55,19 +58,33 @@ def __init__(self, config):
 
         self._last_volume = None
         self._last_mute = None
+        self._observer = None
 
         logger.info(
             f"Mixing using ALSA, {self.device_title}, "
             f"mixer control {self.control!r}."
         )
 
     def on_start(self):
-        self._observer = AlsaMixerObserver(
-            device=self.device,
-            control=self.control,
-            callback=self.actor_ref.proxy().trigger_events_for_changed_values,
+        self._observer = AlsaMixerObserver.start(
+            self._await_mixer(), self.actor_ref.proxy()
+        )
+
+    def on_stop(self):
+        self._stop_observer()
+
+    def on_failure(self, exception_type, exception_value, traceback):
+        self._stop_observer()
+
+    def restart_observer(self, exc=None):
+        self._stop_observer()
+        self._observer = AlsaMixerObserver.start(
+            self._await_mixer(exc), self.actor_ref.proxy()
         )
```

Comment:
It came to my mind, but there's more than one way to implement `_start_observer`:

```python
def _start_observer(self, exc):
    # Perhaps we should raise an exception is observer is already running?
    self._observer = AlsaMixerObserver.start(exc, self.actor_ref.proxy())
```

```python
def _start_observer(self, exc):
    # This check is meaningless iif _start_observer()
    # is used only within AlsaMixer class
    if self._observer is None:
        self._observer = AlsaMixerObserver.start(exc, self.actor_ref.proxy())
```

```python
def _start_observer(self, exc):
    # This is actually restart_observer()
    self._stop_observer()
    self._observer = AlsaMixerObserver.start(exc, self.actor_ref.proxy())
```

I struggled to decide which one to choose, so being guided by "There should be one -- and preferably only one -- obvious way to do it" I introduced `restart_observer()` as public method. This way `_start_observer()` would then need to be implemented as no. 1 from above, which is very repetitive in my opinion and unnecessary because it's internal method.

Does such reasoning make sense or it's still better to explicitly define `_start_observer()` as in no. 1 only for the sake of consistency?

Reply:
I still think it's important for consistency. It helps define the One True Way of doing what needs to be done.